backend/server.js

```diff
@@ -15,15 +52,93 @@ const port = process.env.PORT || 8080
 const app = express()
 
 // Add middlewares to enable cors and json body parsing
+
+const authenticateUser = async (req, res, next) => {
+  try {
+    const user = await User.findOne({ accessToken: req.header('Authorization') })
+    if (user) {
+      req.user = user
+      next()
+    } else {
+      res.status(401).json({ loggedIn: false })
+    }
+  } catch (err) {
+    res.status(403).json({ message: 'Access token missing or invalid', errors: err.errors })
+  }
+}
+
 app.use(cors())
 app.use(bodyParser.json())
 
 // Start defining your routes here
 app.get('/', (req, res) => {
-  res.send('Hello world')
+  res.send('Tiago was here!and Ivett too!!')
+})
+// Signup post
+app.post('/users', async (req, res) => {
+  console.log('ivett')
+  // try to register the user
+  try {
+    const { name, email, password } = req.body
+    if (email.match(/^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/)) {
+      // it is very important to encrypt the passwords and store them encrypted in our db!
+      const user = new User({ name, email, password: bcrypt.hashSync(password) })
+      await user.save()
+      res.status(201).json({ id: user._id, name: user.name, accessToken: user.accessToken })
+      console.log({ user })
+      // if the user is not registered, then we catch the error
+    } else res.status(400).json({ message: 'Invalid email', errors: err.errors })
+  } catch (err) {
+    res.status(400).json({ message: 'Could not create user', errors: err.errors })
+  }
+})
+
+// authentication access point (login/sign in)
+app.post('/sessions', async (req, res) => {
+  try {
+    const user = await User.findOne({ email: req.body.email })
+    if (user && bcrypt.compareSync(req.body.password, user.password)) {
+      //Successful
+      res.status(200).json({ userId: user._id, accessToken: user.accessToken, name: user.name, loggedIn: true })
+    } else {
+      //Failure:
+      // a) user does not exist
+      // b) the encrypted password does not match
+      if (user === null)
+        res.status(404).json({ notFound: true })
+      else res.status(401).json({ message: 'Username or password are incorrect' })
```

Comment:
In a login API, it might be dangerous to return different status codes/responses depending on which information way correct.  

In this case, I could send a bunch of random usernames and when your backend returns 401, I know that I've found a user which exists in your database 🤔 

So we usually stick to 404 - User Not Found, even if the user was correct, but the password was wrong